gossip/src/sync.rs

```diff
@@ -0,0 +1,151 @@
+use std::{
+    collections::VecDeque,
+    sync::{
+        Arc, Condvar, LockResult, Mutex, PoisonError, RwLockReadGuard, RwLockWriteGuard,
+        TryLockError, TryLockResult,
+    },
+};
+
+/// FIFO reader-writer lock.
+#[derive(Default)]
+pub struct RwLock<T> {
+    inner: std::sync::RwLock<T>,
+    queue: Mutex<VecDeque<Arc<Condvar>>>,
+}
+
+impl<T> RwLock<T> {
+    pub fn new(t: T) -> Self {
+        Self {
+            inner: std::sync::RwLock::new(t),
+            queue: Mutex::default(),
+        }
+    }
+
+    pub fn read(&self) -> LockResult<RwLockReadGuard<'_, T>> {
+        use std::sync::RwLock;
+        self.lock(RwLock::read, RwLock::try_read)
+    }
+
+    pub fn write(&self) -> LockResult<RwLockWriteGuard<'_, T>> {
+        use std::sync::RwLock;
+        self.lock(RwLock::write, RwLock::try_write)
+    }
+
+    fn lock<'a, F1, F2, R: 'a>(&'a self, lock: F1, try_lock: F2) -> LockResult<R>
+    where
+        F1: Fn(&'a std::sync::RwLock<T>) -> LockResult<R>,
+        F2: Fn(&'a std::sync::RwLock<T>) -> TryLockResult<R>,
+    {
+        let mut queue = match self.queue.lock() {
+            Ok(queue) => queue,
+            Err(_) => {
+                let guard = lock(&self.inner);
+                return guard.and_then(|r| Err(PoisonError::new(r)));
+            }
+        };
+        // If the queue is empty and no one is waiting, then try lock the inner
+        // and return if it does not block.
+        if queue.is_empty() {
+            match try_lock(&self.inner) {
+                Ok(inner) => return Ok(inner),
+                Err(TryLockError::Poisoned(err)) => return Err(err),
+                Err(TryLockError::WouldBlock) => (),
+            }
+        };
+        // Add self at the end of the queue and
+        // wait until reach head of the queue.
+        let cvar: Arc<Condvar> = Arc::default();
+        queue.push_back(Arc::clone(&cvar));
+        while !Arc::ptr_eq(&cvar, queue.front().unwrap()) {
+            queue = match cvar.wait(queue) {
+                Ok(queue) => queue,
+                Err(_) => {
+                    let guard = lock(&self.inner);
+                    return guard.and_then(|r| Err(PoisonError::new(r)));
+                }
+            };
+            // Check if it is in fact in front of the queue;
+            // Otherwise this has been a spurious wakeup.
+        }
+        // Should drop queue lock here, so that other threads are not blocked
+        // and are queued in order of their arrival.
+        drop(queue);
+        let guard = lock(&self.inner);
+        // Pop self from the queue, and notify the next one in line that it is
+        // now head of the queue.
+        let cvar: Arc<Condvar> = match self.queue.lock() {
```

Comment:
if multiple readers are queued, would they then be evaluated serially?

Reply:
If `r1, r2, w1, r3` are in the queue (with `r1` being the head of the queue), then `r1` and `r2` will share the lock but `r3` has to wait until `w1` is done because of fifo.
So readers will share the lock as long as there is no writer between them.